discovery_engine/lib/src/domain/search_manager.dart

```diff
@@ -0,0 +1,217 @@
+  final ChangedDocumentRepository _changedRepo;
+  final EngineStateRepository _engineStateRepo;
+
+  SearchManager(
+    this._engine,
+    this._config,
+    this._searchRepo,
+    this._docRepo,
+    this._activeRepo,
+    this._changedRepo,
+    this._engineStateRepo,
+  );
+
+  /// Handle the given search client event.
+  ///
+  /// Fails if [event] does not have a handler implemented.
+  Future<EngineEvent> handleSearchClientEvent(SearchClientEvent event) =>
+      event.maybeWhen(
+        searchRequested: searchRequested,
+        nextSearchBatchRequested: nextSearchBatchRequested,
+        restoreSearchRequested: restoreSearchRequested,
+        searchClosed: searchClosed,
+        orElse: () =>
+            throw UnimplementedError('handler not implemented for $event'),
+      );
+
+  Future<List<api.Document>> _getSearchDocuments(ActiveSearch search) async {
+    final searchDocs =
+        await _engine.getSearchDocuments(search, _config.maxSearchDocs);
+    await _engineStateRepo.save(await _engine.serialize());
+
+    await _docRepo.updateMany(searchDocs.map((e) => e.document));
+    for (final docWithData in searchDocs) {
+      final id = docWithData.document.documentId;
+      await _activeRepo.update(id, docWithData.data);
+    }
+
+    return searchDocs
+        .map((docWithData) => docWithData.document.toApiDocument())
+        .toList();
+  }
+
+  /// Obtain the first batch of search documents and persist to repositories.
+  Future<EngineEvent> searchRequested(
+    String queryTerm,
+    FeedMarket market,
+  ) async {
+    await searchClosed();
+
+    final search = ActiveSearch(
+      queryTerm: queryTerm,
+      requestedPageNb: 1,
+      pageSize: _config.maxSearchDocs,
+      market: market,
+    );
+    final docs = await _getSearchDocuments(search);
+
+    if (docs.isEmpty) {
+      const reason = SearchFailureReason.noResultsAvailable;
+      return const EngineEvent.searchRequestFailed(reason);
+    }
+
+    // TODO: should we persist active search before we get documents?
+    await _searchRepo.save(search);
+
+    return EngineEvent.searchRequestSucceeded(search, docs);
+  }
+
+  /// Obtain the next batch of search documents and persist to repositories.
+  Future<EngineEvent> nextSearchBatchRequested() async {
+    var search = await _searchRepo.getCurrent();
+
+    if (search == null) {
+      const reason = SearchFailureReason.noActiveSearch;
+      return const EngineEvent.nextSearchBatchRequestFailed(reason);
+    }
+
+    // lets update active search params
+    search = search.copyWith(requestedPageNb: search.requestedPageNb + 1);
+    final docs = await _getSearchDocuments(search);
+
+    if (docs.isEmpty) {
+      const reason = SearchFailureReason.noResultsAvailable;
+      return const EngineEvent.nextSearchBatchRequestFailed(reason);
+    }
+
+    // TODO: should we persist updated search params before we get documents?
+    await _searchRepo.save(search);
```

Comment:
Yes, if we get an error back from rust we should try the same page again.

Reply:
so then we should persist the search object after we get documents from rust, so we (the app) can send the same event again, and expect same results (especially for getting next page, the incremented page nb should be updated before we ask engine for next page but saved after we received the next page documents)

Reply:
I've changed that for now, so we persist search before we call the engine, but maybe I should revert it to the prev implementation, where we persisted after calling the engine?

Reply:
it looks like persisting it after the call to the engine would be better. currently if `_getSearchDocuments()` fails and we enter `nextSearchBatchRequested()` again, the counter `search.requestedPageNb` is already incremented but we didn't get the docs of the previous page. this could happen multiple times, which leaves us at later pages with worse results than necessary.